src/datanode/src/datanode.rs

```diff
@@ -490,6 +469,63 @@ impl DatanodeBuilder {
         Ok(engines)
     }
 
+    /// Builds [MitoEngine] according to options.
+    async fn build_mito_engine(
+        opts: &DatanodeOptions,
+        object_store_manager: ObjectStoreManagerRef,
+        config: MitoConfig,
+    ) -> Result<MitoEngine> {
+        // TODO(niebayes): validate wal config, for e.g. providing a remote wal for standalone mode is not allowed.
+
+        let mito_engine = match &opts.wal {
+            WalConfig::RaftEngine(raft_engine_config) => MitoEngine::new(
+                config,
+                Self::build_raft_engine_log_store(&opts.storage.data_home, raft_engine_config)
+                    .await?,
+                object_store_manager,
+            ),
+            WalConfig::Kafka(kafka_config) => MitoEngine::new(
+                config,
+                Self::build_kafka_log_store(kafka_config).await?,
+                object_store_manager,
+            ),
+        };
+        Ok(mito_engine)
+    }
+
+    /// Builds [RaftEngineLogStore].
+    async fn build_raft_engine_log_store(
+        data_home: &str,
+        config: &RaftEngineConfig,
+    ) -> Result<Arc<RaftEngineLogStore>> {
+        let data_home = normalize_dir(data_home);
+        let wal_dir = match &config.dir {
+            Some(dir) => dir.clone(),
+            None => format!("{}{WAL_DIR}", data_home),
```

Comment:
```suggestion
            None => format!("{}{WAL_DIR}", normalize_dir(data_home)),
```

Reply:
The `datanode` was already normalized. See line 501.